include/souper/Extractor/Solver.h

```diff
@@ -41,6 +41,36 @@ class Solver {
                                                    const std::vector<InstMapping> &PCs,
                                                    Inst *LHS, InstContext &IC) = 0;
   virtual std::string getName() = 0;
+  virtual
+  std::error_code negative(const BlockPCs &BPCs,
+                           const std::vector<InstMapping> &PCs,
+                           Inst *LHS, llvm::APInt &Negative,
+                           InstContext &IC) = 0;
+  virtual
+  std::error_code knownBits(const BlockPCs &BPCs,
+                            const std::vector<InstMapping> &PCs,
+                            Inst *LHS, llvm::APInt &Zeros, llvm::APInt &Ones,
+                            InstContext &IC) = 0;
+  virtual
+  std::error_code nonNegative(const BlockPCs &BPCs,
+                            const std::vector<InstMapping> &PCs,
+                            Inst *LHS, llvm::APInt &NonNegative,
```

Comment:
I don't think it's necessary to allocate an APInt for facts like NonNegative, this should just be a bool